xl-v1-createfile.go

```diff
@@ -0,0 +1,346 @@
+/*
+ * Minio Cloud Storage, (C) 2016 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"hash"
+	"io"
+	slashpath "path"
+	"strconv"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	fastSha512 "github.com/minio/minio/pkg/crypto/sha512"
+)
+
+// Erasure block size.
+const erasureBlockSize = 4 * 1024 * 1024 // 4MiB.
+
+// cleanupCreateFileOps - cleans up all the temporary files and other
+// temporary data upon any failure.
+func (xl XL) cleanupCreateFileOps(volume, path string, writers ...io.WriteCloser) {
+	closeAndRemoveWriters(writers...)
+	for _, disk := range xl.storageDisks {
+		disk.DeleteFile(volume, path)
+	}
+}
+
+// Close and remove writers if they are safeFile.
+func closeAndRemoveWriters(writers ...io.WriteCloser) {
+	for _, writer := range writers {
+		safeCloseAndRemove(writer)
+	}
+}
+
+type quorumDisk struct {
+	disk  StorageAPI
+	index int
+}
+
+// getQuorumDisks - get the current quorum disks.
+func (xl XL) getQuorumDisks(volume, path string) (quorumDisks []quorumDisk, higherVersion int64) {
+	fileQuorumVersionMap := xl.getFileQuorumVersionMap(volume, path)
+	for diskIndex, formatVersion := range fileQuorumVersionMap {
+		if formatVersion > higherVersion {
+			higherVersion = formatVersion
+			quorumDisks = []quorumDisk{{
+				disk:  xl.storageDisks[diskIndex],
+				index: diskIndex,
+			}}
+
+		} else if formatVersion == higherVersion {
+			quorumDisks = append(quorumDisks, quorumDisk{
+				disk:  xl.storageDisks[diskIndex],
+				index: diskIndex,
+			})
+
+		}
+	}
+	return quorumDisks, higherVersion
+}
+
+func (xl XL) getFileQuorumVersionMap(volume, path string) map[int]int64 {
+	metadataFilePath := slashpath.Join(path, metadataFile)
```

Comment:
This function is redundant. We only need one function that returns meta-data of all quorum'd disks. Then loop thru the results to peek into values.